ekncontent/ekncontent/eknc-domain.c

```diff
@@ -739,40 +739,111 @@ static void
+  ScatterTaskCountdownCompleteCallback callback;
+  gpointer data;
+  GError *error;
+} ScatterTaskCountdown;
+
+static ScatterTaskCountdown *
+scatter_task_countdown_new (guint remaining,
+                            ScatterTaskCountdownCompleteCallback callback,
+                            gpointer data)
+{
+  ScatterTaskCountdown *countdown = g_new0 (ScatterTaskCountdown, 1);
+  countdown->remaining = remaining;
+  countdown->callback = callback;
+  countdown->data = data;
+
+  return countdown;
+}
+
+static void
+scatter_task_countdown_subtask_completed (ScatterTaskCountdown **countdown_ptr,
+                                          GError *error)
+{
+  ScatterTaskCountdown *countdown = *countdown_ptr;
+
+  /* If we received an error we should still continue to countdown, but
+   * store the error so that the caller can deal with it */
+  if (!countdown->error)
+    countdown->error = error;
+  else
+    g_clear_error (&error);
+
+  if (--countdown->remaining == 0)
+    {
+      (*countdown->callback) (error, countdown->data);
+      g_free (countdown);
+      *countdown_ptr = NULL;
+    }
+}
+
+typedef struct {
+  /* Note that scatter_task_countdown_subtask_completed will free
+   * this member once the countdown has completed */
+  ScatterTaskCountdown *countdown;
+  EkncDomain *domain;
+  GPtrArray *models;
+  guint index;
+} ObjectResponseData;
+
+static ObjectResponseData *
+object_response_data_new (ScatterTaskCountdown *countdown,
+                          EkncDomain *domain,
+                          GPtrArray *models,
+                          guint index)
+{
+  ObjectResponseData *data = g_new0 (ObjectResponseData, 1);
+  data->countdown = countdown;
+  data->domain = domain;
+  data->models = models;
+  data->index = index;
+  return data;
+}
+
+static void
+on_received_all_model_objects (GError *error, gpointer user_data)
+{
+  g_autoptr(GTask) task = user_data;
+
+  /* Received an error, propogate it to the caller */
+  if (error)
+    g_task_return_error (task, error);
+  else
+    g_task_return_boolean (task, TRUE);
+}
+
+static void
 on_object_response (GObject *source,
                     GAsyncResult *result,
                     gpointer user_data)
 {
-  g_autoptr(GTask) task = user_data;
-  EkncDomain *domain = g_task_get_source_object (task);
-  QueryState *state = g_task_get_task_data (task);
+  g_autofree ObjectResponseData *object_response_data = user_data;
+  EkncDomain *domain = object_response_data->domain;
   GError *error = NULL;
 
-  if (g_task_get_completed (task))
-    return;
-
   EkncContentObjectModel *model = eknc_domain_get_object_finish (domain, result, &error);
-  if (error != NULL)
-    {
-      g_task_return_error (task, error);
-      return;
-    }
 
-  state->models = g_slist_append (state->models, model);
+  *(&g_ptr_array_index (object_response_data->models, object_response_data->index)) = model;
+  scatter_task_countdown_subtask_completed (&object_response_data->countdown, error);
```

Comment:
I think the error ownership around this function could be improved; `scatter_task_countdown_subtask_completed()` should make a copy of the error when needed, and this function should declare the error with `g_autoptr()`